osu.Game.Rulesets.Mania/Objects/Drawables/DrawableNote.cs

```diff
@@ -89,17 +95,24 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
             if (!userTriggered)
             {
                 if (!HitObject.HitWindows.CanBeHit(timeOffset))
+                {
+                    bonusNote!.TriggerResult(false);
                     ApplyResult(r => r.Type = r.Judgement.MinResult);
+                }
+
                 return;
             }
 
             var result = HitObject.HitWindows.ResultFor(timeOffset);
             if (result == HitResult.None)
                 return;
 
-            ApplyResult(r => r.Type = result);
+            bonusNote!.TriggerResult(result == HitResult.Perfect);
+            ApplyResult(r => r.Type = MutateResultApplication(result));
         }
 
+        protected virtual HitResult MutateResultApplication(HitResult result) => result;
```

Comment:
I can't say I like this virtual too much...

The counterproposal here would be something along the lines of (warning, untested):

```diff
diff --git a/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNoteTail.cs b/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNoteTail.cs
index a183231310..24b028818e 100644
--- a/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNoteTail.cs
+++ b/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNoteTail.cs
@@ -36,15 +36,20 @@ public DrawableHoldNoteTail(TailNote tailNote)
             // Factor in the release lenience
             base.CheckForResult(userTriggered, timeOffset / TailNote.RELEASE_WINDOW_LENIENCE);
 
-        protected override HitResult MutateResultApplication(HitResult result)
+        protected override HitResult? CheckForMainNoteResult(bool userTriggered, double timeOffset)
         {
+            var baseResult = base.CheckForMainNoteResult(userTriggered, timeOffset);
+
+            if (baseResult == null)
+                return null;
+
             // If the head wasn't hit or the hold note was broken, cap the max score to Meh.
             bool hasComboBreak = !HoldNote.Head.IsHit || HoldNote.Body.HasHoldBreak;
 
-            if (result > HitResult.Meh && hasComboBreak)
+            if (baseResult > HitResult.Meh && hasComboBreak)
                 return HitResult.Meh;
 
-            return result;
+            return baseResult;
         }
 
         public override bool OnPressed(KeyBindingPressEvent<ManiaAction> e) => false; // Handled by the hold note
diff --git a/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableNote.cs b/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableNote.cs
index eaeac33a11..a37fc8ba8d 100644
--- a/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableNote.cs
+++ b/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableNote.cs
@@ -89,6 +89,17 @@ protected override void OnDirectionChanged(ValueChangedEvent<ScrollingDirection>
         }
 
         protected override void CheckForResult(bool userTriggered, double timeOffset)
+        {
+            var result = CheckForMainNoteResult(userTriggered, timeOffset);
+
+            if (result != null)
+            {
+                bonusNote!.TriggerResult(result == HitResult.Perfect);
+                ApplyResult(r => r.Type = result.Value);
+            }
+        }
+
+        protected virtual HitResult? CheckForMainNoteResult(bool userTriggered, double timeOffset)
         {
             Debug.Assert(HitObject.HitWindows != null);
 
@@ -97,22 +108,15 @@ protected override void CheckForResult(bool userTriggered, double timeOffset)
                 if (!HitObject.HitWindows.CanBeHit(timeOffset))
                 {
                     bonusNote!.TriggerResult(false);
-                    ApplyResult(r => r.Type = r.Judgement.MinResult);
+                    return Result.Judgement.MinResult;
                 }
 
-                return;
+                return null;
             }
 
-            var result = HitObject.HitWindows.ResultFor(timeOffset);
-            if (result == HitResult.None)
-                return;
-
-            bonusNote!.TriggerResult(result == HitResult.Perfect);
-            ApplyResult(r => r.Type = MutateResultApplication(result));
+            return HitObject.HitWindows.ResultFor(timeOffset);
         }
 
-        protected virtual HitResult MutateResultApplication(HitResult result) => result;
-
         public virtual bool OnPressed(KeyBindingPressEvent<ManiaAction> e)
         {
             if (e.Action != Action.Value)

```

Reply:
I unfortunately find your version even harder to follow. Also I don't think it's correct logic-side due to the `userTriggered` case not being correctly considered.

I've made a minor refactor to what I had, can you see if it reads better? Of note, it's likely more correct now as previously the perfect bonus could be given for a "Meh" result.

Reply:
Yeah I dunno... My primary (weak and argumentative) problem I guess is that I'd consider the current method as a "weak" virtual method, by which I mean that it was created in the most straight-line way possible to fix an inheritance issue and this type of design backfires pretty hard on almost any sort of future extension. Then again, I'm not sure we're ever gonna have more object types in mania so probably fine...

Reply:
I guess my only alternative would be code duplication. But the reason I changed this in the first place was due to a regression from not duplicating code correctly 😓

Reply:
Yeah I guess the current version is somewhat better than duplication.

Probably fine to keep as is for now.